File: 2022/16.py
from functools import cache, reduce
import re


def get_input(puzzle: str) -> tuple[list[int], list[set[int]]]:
    n = len(puzzle.splitlines())
    graph = [set() for _ in range(n)]
    rates = [0] * n
    names = {"AA": 0}

    def name(n: str) -> int:
        return names.setdefault(n, len(names))

    for line in puzzle.splitlines():
        v, r, n = re.match(r"^Valve (.+) has flow rate=(\d+); tunnels? leads? to valves? (.+)$", line).groups()  # type: ignore
        v = name(v)
        r = int(r)
        n = list(map(name, n.split(", ")))
        rates[v] = r
        graph[v].update(n)
    return rates, graph


def part1(puzzle: str):
    rates, graph = get_input(puzzle)
    n = len(rates)
    dist = {i: {j: 0 if i == j else 1 if j in graph[i] else 1e1337 for j in range(n)} for i in range(n)}
    for k in range(n):
        for i in range(n):
            for j in range(n):
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

    @cache
    def solve(p, time, closed):
        if time <= 0:
            return 0

        out = 0
        for q in range(n):
            if closed & 1 << q == 0:
                continue
            t = time - dist[p][q] - 1
            out = max(out, solve(q, t, closed & ~(1 << q)) + rates[q] * t)
        return out

    return solve(0, 30, reduce(lambda acc, x: acc | 1 << x, (i for i in range(n) if rates[i]), 0))


def part2(puzzle: str):
    rates, graph = get_input(puzzle)
    n = len(rates)
    dist = {i: {j: 0 if i == j else 1 if j in graph[i] else 1e1337 for j in range(n)} for i in range(n)}
    for k in range(n):
        for i in range(n):
            for j in range(n):
                dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

    @cache
    def solve(p, time, closed):
        if time <= 0:
            return 0

        out = 0
        for q in range(n):
            if closed & 1 << q == 0:
                continue
            t = time - dist[p][q] - 1
            out = max(out, solve(q, t, closed & ~(1 << q)) + rates[q] * t)
        return out

    out = 0
    valves = [i for i in range(n) if rates[i]]
    for s in range(1 << len(valves)):
        a = solve(0, 26, reduce(lambda acc, x: acc | 1 << x, (j for i, j in enumerate(valves) if s & 1 << i), 0))
        b = solve(0, 26, reduce(lambda acc, x: acc | 1 << x, (j for i, j in enumerate(valves) if s & 1 << i == 0), 0))
        out = max(out, a + b)
    return out


if __name__ == "__main__":
    from aoc import run

    run(2022, 16, part1, part2)

File: 2022/16.rs
use regex::Regex;
use rustc_hash::FxHashMap;

type Input = Vec<Valve>;

struct Valve {
    rate: u32,
    tunnels: Vec<usize>,
}

fn setup(input: &str) -> Input {
    let regex = Regex::new(r"^Valve ([A-Z]+) .* rate=(\d+); .* valves? ([A-Z, ]+)$").unwrap();
    let mut names = FxHashMap::default();
    names.insert("AA", 0);
    let mut name = |n| {
        let cnt = names.len();
        *names.entry(n).or_insert(cnt)
    };

    let mut valves: FxHashMap<usize, Valve> = input
        .trim()
        .lines()
        .map(|line| {
            let caps = regex.captures(line).unwrap();
            let v = name(caps.get(1).unwrap().as_str());
            let rate = caps[2].parse().unwrap();
            let tunnels = caps
                .get(3)
                .unwrap()
                .as_str()
                .split(", ")
                .map(&mut name)
                .collect();
            (v, Valve { rate, tunnels })
        })
        .collect();

    (0..valves.len())
        .map(|x| valves.remove(&x).unwrap())
        .collect()
}

struct Solver<'a> {
    valves: &'a Input,
    dist: Vec<Vec<u32>>,
    dp: FxHashMap<(usize, u32, u64), u32>,
}

impl<'a> Solver<'a> {
    fn new(valves: &'a Input) -> Self {
        let mut dist = vec![vec![u32::MAX; valves.len()]; valves.len()];
        for (i, v) in valves.iter().enumerate() {
            dist[i][i] = 0;
            v.tunnels.iter().for_each(|&j| dist[i][j] = 1);
        }
        for k in 0..valves.len() {
            for i in 0..valves.len() {
                for j in 0..valves.len() {
                    if let Some(d) = dist[i][k].checked_add(dist[k][j]) {
                        dist[i][j] = dist[i][j].min(d);
                    }
                }
            }
        }

        Self {
            valves,
            dist,
            dp: FxHashMap::default(),
        }
    }

    fn solve(&mut self, p: usize, time: u32, closed: u64) -> u32 {
        let key = (p, time, closed);
        if let Some(&result) = self.dp.get(&key) {
            return result;
        }

        let result = (0..self.valves.len())
            .filter_map(|q| {
                if closed & 1 << q == 0 {
                    return None;
                }
                if let Some(t) = time.checked_sub(self.dist[p][q] + 1) {
                    Some(self.solve(q, t, closed & !(1 << q)) + self.valves[q].rate * t)
                } else {
                    None
                }
            })
            .max()
            .unwrap_or(0);
        self.dp.insert(key, result);
        result
    }
}

fn part1(input: &Input) -> u32 {
    Solver::new(input).solve(
        0,
        30,
        input
            .iter()
            .enumerate()
            .filter(|(_, x)| x.rate > 0)
            .fold(0, |acc, (i, _)| acc | 1 << i),
    )
}

fn part2(input: &Input) -> u32 {
    let mut solver = Solver::new(input);
    let valves = input
        .iter()
        .enumerate()
        .filter(|(_, x)| x.rate > 0)
        .map(|(i, _)| i)
        .collect::<Vec<_>>();
    (0u64..1 << (valves.len() - 1))
        .filter(|&s| valves.len().abs_diff(s.count_ones() as usize * 2) <= 1)
        .map(|s| {
            let a = solver.solve(
                0,
                26,
                valves
                    .iter()
                    .enumerate()
                    .filter(|&(i, _)| s & 1 << i != 0)
                    .fold(0, |acc, (_, j)| acc | 1 << j),
            );
            let b = solver.solve(
                0,
                26,
                valves
                    .iter()
                    .enumerate()
                    .filter(|&(i, _)| s & 1 << i == 0)
                    .fold(0, |acc, (_, j)| acc | 1 << j),
            );
            a + b
        })
        .max()
        .unwrap()
}

aoc::main!(2022, 16);
aoc::example!(ex01, "16.1.txt", 1651, 1707);
aoc::test_input!("16.txt", 1991, 2705);
